File: 2024/46/lan.rb
#!/usr/bin/env ruby

# 2024 Day 23 Part 2 / Puzzle 2024/46

# Something about finding sets of connected computers

class Puzzle
  def initialize()
    @codes = [ ]
    @conns = { }
    @edges = { }
  end

  def decode(line)
    l, r = line.split('-')

    if !@conns.has_key? (l)
      @conns[l] = [ ]
    end

    if !@conns.has_key? (r)
      @conns[r] = [ ]
    end

    @conns[l].push(r)
    @conns[r].push(l)
    @edges[[l,r].sort.join] = 1
  end

  def connections
    @conns
  end

  def all_connected?(*ns)
    ns.combination(2).map { |comb|
      comb.sort.join
    }.all? { |key| @edges.has_key?(key) }
  end

end

p = Puzzle.new()

File.foreach(ARGV.shift || '../45/input') { |line|
  p.decode(line.chomp)
}

visited = { }

largest_match = ''
warns = 0

found = false
num_combos = 14

while !found
  puts "Running through combinations of #{num_combos}"

  p.connections.each_pair { |n, conn|
    if num_combos > conn.length
      break
    end

    conn.combination(num_combos) { |comb|
      all_n = [n, *comb]

      key = all_n.sort.join (',')

      if !visited.has_key?(key)
        visited[key] = 1

        if p.all_connected?(*all_n)
          puts "#{key} success"

          if key.length > largest_match.length
            largest_match = key
            found = true
          elsif key.length == largest_match.length
            puts "WARN #{key} is a duplicate match!"
            warns += 1
          end
        end
      end
    }
  }

  if !found
    num_combos -= 1
    if num_combos <= 10
      break;
    end
  end
end

puts "Largest match: #{largest_match} (#{warns} warnings)"

File: 2024/46/lan.rs
use std::collections::HashMap;
use std::env;
use std::fs;

// Advent of Code: 2024 day 23, part 2

type Graph = HashMap<String, Vec<String>>;

fn is_connected(g: &Graph, n1: &str, n2: &str) -> bool {
    let (a, b) = if n1 < n2 { (n1, n2) } else { (n2, n1) };

    return match g.get(a) {
        Some(list) => list.iter().any(|x| x == b),
        None => false,
    };
}

fn all_connected(g: &Graph, nodes: &[&str]) -> bool {
    // base case is if there are two nodes
    if nodes.len() == 2 {
        return is_connected(g, nodes[0], nodes[1]);
    }

    // Otherwise bite off a node and then ensure the rest are connected to each other, and to the
    // node

    let (n, rest) = (nodes[0], &nodes[1..]);
    if !all_connected(g, rest) {
        return false;
    }

    return rest.iter().all(|x| is_connected(g, n, x));
}

fn read_graph(lines: Vec<String>) -> Graph {
    let mut graph = Graph::new();

    for line in lines {
        let l = String::from(&line[0..2]);
        let r = String::from(&line[3..5]);

        // The one < the other will be the key, the other will be in the value
        // (the list of the key's connections)
        let (a, b) = if l < r { (l, r) } else { (r, l) };

        let deps = graph.entry(a).or_insert(Vec::<String>::new());
        (*deps).push(b);
    }

    for (_, list) in graph.iter_mut() {
        list.sort();
    }

    return graph;
}

fn combinations_of<'a>(n: usize, v: &'a [&'a str]) -> Vec<Vec<&'a str>> {
    let mut res: Vec<Vec<&str>> = vec![];

    // base case
    if n == 1 {
        for s in v {
            let combo_list: Vec<&str> = vec![s];
            res.push(combo_list);
        }

        return res;
    }

    // handle our one node by pushing to the end of every array
    // in the list we get
    for i in 0..=(v.len()-n) {
        let mut inner_res = combinations_of(n - 1, &v[(i+1)..]);
        for list in inner_res.iter_mut() {
            list.push(v[i]);
        }

        res.append(&mut inner_res);
    }

    return res;
}

fn main() {
    let default_filename: &'static str = "../45/input";
    let args: Vec<String> = env::args().collect();
    let in_file: String = match args.iter().filter(|a| !a.starts_with("-")).count() {
        // No args provided save argv[0]
        1 => String::from(default_filename),
        _ => args[args.len() - 1].clone(),
    };

    let lines = fs::read_to_string(&in_file)
        .expect("Should have been able to read the file")
        .lines()
        .map(String::from)
        .collect::<Vec<_>>();

    let g = read_graph(lines);

    // Loop through progressively smaller combination counts until we find the first (and hopefully
    // only) clique, which will be the largest.
    let mut num_combos = g.values().map(|x| x.len()).max().unwrap();

    loop {
        for (n1, conns) in g.iter() {
            let mut nodes: Vec<&str> = vec![&n1];
            nodes.extend(conns.iter().map(|x| &x[..]));

            // Not every node is connected to lots of others
            if num_combos >= nodes.len() {
                continue;
            }

            for c in combinations_of(num_combos, &nodes) {
                if all_connected(&g, &c) {
                    let mut res = c.clone();
                    res.sort();

                    println!("Found a match with {}", res.join(","));
                    return;
                }
            }
        }

        num_combos -= 1;
        if num_combos < 10 {
            println!("Something went wrong I guess!");
            return;
        }
    }
}
